### PR TITLE
 List: Added scrolling mode for scrollToIndex

### LINKED ISSUE #2996: List: Improve the scrollToIndex functionality
In the current implementation the `List` ensures that the element provided by the index is visible but doesn't support alignment of the element within the visible window. 

My usecase is that I have a list with group headers and when clicking on the group header the user is presented with a menu where he can scroll to another group header. If possible this new group header should then be the first entry in the visible window. 

`react-virtualized` has `scrollToAlignment` on its list that fits exactly my usecase: 
image

### PR DESCRIPTION
**Pull request by lambertwang**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #2996
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Added scrolling modes to list

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/List/List.tsx [hunk 1]
```diff
@@ -207,50 +213,65 @@ export class List extends BaseComponent<IListProps, IListState> implements IList
...
+          for (let itemIndexInPage = 0; itemIndexInPage < itemPositionWithinPage; ++itemIndexInPage) {
+            scrollTop += measureItem(itemIndex + itemIndexInPage);
+          }
+          const scrollBottom = scrollTop + measureItem(index);
+
           const itemIsFullyVisible = scrollTop >= scrollWindow.top && scrollBottom <= scrollWindow.bottom;
-          if (itemIsFullyVisible) {
-            // Item is already visible, do nothing.
+          // Item is already visible and scrollMode is auto, do nothing and return
+          if (itemIsFullyVisible && scrollToMode === ScrollToMode.auto) {
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/List/List.tsx]: what if `scrollToMode` is undefined?

File: packages/office-ui-fabric-react/src/components/List/List.tsx [hunk 2]
```diff
@@ -207,50 +213,65 @@ export class List extends BaseComponent<IListProps, IListState> implements IList
...
             return;
           }
 
           const itemIsPartiallyAbove = scrollTop < scrollWindow.top;
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/List/List.tsx]: I don't like how the new options are used with the old logic, can we restructure to make that clearer?

Comment (cschleiden) [reply packages/office-ui-fabric-react/src/components/List/List.tsx]: Why calculate the `itemIsPartially...` values at all if your scrollToMode option is set?

File: packages/office-ui-fabric-react/src/components/List/List.tsx (hunk 1 above)
Comment (lambertwang) [reply packages/office-ui-fabric-react/src/components/List/List.tsx]: It is defaulted to auto

File: packages/office-ui-fabric-react/src/components/List/List.types.ts [hunk 3]
```diff
@@ -2,6 +2,21 @@ import * as React from 'react';
...
+export const enum ScrollToMode {
+  /**
+   * Does not make any consideration to where in the viewport the item should align to.
+   */
+  auto,
+  /**
+   * Attempts to scroll the list so the top of the desired item is aligned with the top of the viewport.
+   */
+  top,
+  /**
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/List/List.types.ts]: Just curious, why wouldn't you just always want the behavior to be:

Scroll the item to the center of the viewport?

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/List/List.types.ts]: Actually in reading the issue and the react-virtualized approach, I'm fine with the change, but it seems that center is missing...

Comment (lambertwang) [reply packages/office-ui-fabric-react/src/components/List/List.types.ts]: Ok, I can add center.

Comment (lambertwang) [reply packages/office-ui-fabric-react/src/components/List/List.types.ts]: Done.

**Comment by lambertwang** [comment]
@cschleiden @aditima Can I get a review?

**Comment by Jahnp** [comment]
Checking in during triage to make sure there's traction. @cschleiden , @aditima , would you be able to take some time this week to review? Thank you!

File: packages/office-ui-fabric-react/src/components/List/examples/List.Scrolling.Example.tsx [hunk 4]
```diff
@@ -1,13 +1,15 @@
 import * as React from 'react';
 import {
+  autobind,
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/List/examples/List.Scrolling.Example.tsx]: no need to use autobind anymore.

File: packages/office-ui-fabric-react/src/components/List/List.tsx [hunk 5]
```diff
@@ -468,7 +492,7 @@ export class List extends BaseComponent<IListProps, IListState> implements IList
     return (
       <div { ...divProps }>
         { cells }
-      </div>
+      </div >
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/List/List.tsx]: nit, undo

**Pull request merged**